### ISSUE TITLE
Returning complex types from objmode fails in some cases

### ISSUE THREAD

**Issue by kozlov-alexey** [numba/numba#6668]
- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Hi,

I tested #4769 and found some issues with it. Please take a look at below reproducers.

Use-case 1: works OK with ret type provided as global variable, but doesn't work with concrete types
```
import numba
from numba import objmode, typeof, njit
from numba.core import types

'''
    raise errors.CompilerError(
    numba.core.errors.CompilerError: Failed in nopython mode pipeline (step: Handle with contexts)
    objectmode context requires constants string for type annotation
'''
@njit
def test1():
    with objmode(res=types.int32):      # using concrete type is not permitted
        res = 32                        # raised error still refers to old requirement
    return res
 
result = test1()
print(result)
```

Use-case 2: returning python list from objmode.
```
'''
    Gives:
    ret_type=reflected list(int64)<iv=None>
    [<NULL>, 2, 3, 4, 5]
'''
ret_type = typeof([1, 2, 3, 4, 5])      # NOK (garbage returned)
print(f"ret_type={ret_type}")
@njit
def test2():
    with objmode(res=ret_type):
        res = [1, 2, 3, 4, 5]
    return res
 
result = test2()
print(result)
```

Use-case 3: returning python set from objmode.
```
'''
    Gives:
    File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\envs\NUMBA_MASTER\lib\site-packages\llvmlite\binding\module.py", line 25, in parse_assembly
        raise RuntimeError("LLVM IR parsing error\n{0}".format(errmsg))
    RuntimeError: Failed in nopython mode pipeline (step: Handle with contexts)
    Failed in nopython mode pipeline (step: nopython mode backend)
    LLVM IR parsing error
    <string>:410:7: error: value doesn't match function result type 'i32'
      ret i8* null
          ^
'''
ret_type = typeof({1, 2, 3, 4, 5})      # NOK (LLVM parsing error)
print(f"ret_type={ret_type}")
@njit
def test3():
    with objmode(res=ret_type):
        res = {1, 2, 3, 4, 5}
    return res

result = test3()
print(result)
```

Use-case 4: returning python dict from objmode.
```
ret_type = types.DictType(types.unicode_type, types.int64)      # NOK (segfaults)
print(f"ret_type={ret_type}")
@njit
def test4():
    with objmode(res=ret_type):
        res = {'A': 1, 'B': 2}
    return res

result = test4()
print(result)
```

Kind Regards,
Alexey.

**Comment by esc** [comment]
@kozlov-alexey thank you for submitting this! I've marked it as needing a bit of triage and hope that someone can look at it shortly.

**Comment by stuartarchibald** [comment]
Hi @kozlov-alexey, thanks for the report(s)!

---------------
> Use-case 1: works OK with ret type provided as global variable, but doesn't work with concrete > types

This needs looking at, I think `numba.types.*` needs to be added to the legal argument kinds in the `_ObjModeContextType` with-lifting, and evidently the error message needs an update!

---------------
Use-case 2: returning python list from objmode.
```
'''
    Gives:
    ret_type=reflected list(int64)<iv=None>
    [<NULL>, 2, 3, 4, 5]
'''
ret_type = typeof([1, 2, 3, 4, 5])      # NOK (garbage returned)
print(f"ret_type={ret_type}")
@njit
def test2():
    with objmode(res=ret_type):
        res = [1, 2, 3, 4, 5]
    return res
 
result = test2()
print(result)
```

I think the issue here is that `ret_type` has the property `reflected` by virtue of using `typeof` on a python list, whereas the type escaping the `objmode` block is not reflected. This works:
```python
from numba import njit, typeof, objmode, types
import numpy as np

ret_type = types.List(types.int64, reflected=False)
print(f"ret_type={ret_type}")
@njit
def test2():
    with objmode(res=ret_type):
        res = [1, 2, 3, 4, 5]
    return res

result = test2()
print(result)
```

------------
Use-case 3: returning python set from objmode.
```
'''
    Gives:
    File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\envs\NUMBA_MASTER\lib\site-packages\llvmlite\binding\module.py", line 25, in parse_assembly
        raise RuntimeError("LLVM IR parsing error\n{0}".format(errmsg))
    RuntimeError: Failed in nopython mode pipeline (step: Handle with contexts)
    Failed in nopython mode pipeline (step: nopython mode backend)
    LLVM IR parsing error
    <string>:410:7: error: value doesn't match function result type 'i32'
      ret i8* null
          ^
'''
ret_type = typeof({1, 2, 3, 4, 5})      # NOK (LLVM parsing error)
print(f"ret_type={ret_type}")
@njit
def test3():
    with objmode(res=ret_type):
        res = {1, 2, 3, 4, 5}
    return res

result = test3()
print(result)
```

I think the issue here is quite possible down to some issue with call conventions and error handling, this is a guess based on what's in the IR:
```llvm
  %".220" = call i32 @"_ZN5numba7cpython6setobj11SetInstance21_pyapi_get_hash_value12$3clocals$3e11wrapper$247Ex"(i64* %".216", {i8*, i32, i8*}** %"excinfo.1", i64 %".205")
  %".221" = load {i8*, i32, i8*}*, {i8*, i32, i8*}** %"excinfo.1"
  %".222" = icmp eq i32 %".220", 0
  %".223" = icmp eq i32 %".220", -2
  %".224" = icmp eq i32 %".220", -1
  %".225" = icmp eq i32 %".220", -3
  %".226" = or i1 %".222", %".223"
  %".227" = xor i1 %".226", -1
  %".228" = icmp sge i32 %".220", 1
  %".229" = select i1 %".228", {i8*, i32, i8*}* %".221", {i8*, i32, i8*}* undef
  %".230" = load i64, i64* %".216"
  store i1 0, i1* %".231"
  store i64 0, i64* %".234"
  br i1 %".227", label %"bb_body.endif.endif.endif.endif.if", label %"bb_body.endif.endif.endif.endif.else"
bb_body.endif.endif.endif.endif.if:
  store i1 %".227", i1* %".231"
  br i1 %".228", label %"bb_body.endif.endif.endif.endif.if.if", label %"bb_body.endif.endif.endif.endif.if.endif"
bb_body.endif.endif.endif.endif.else:
  store i64 %".230", i64* %".234"
  br label %"bb_body.endif.endif.endif.endif.endif"
bb_body.endif.endif.endif.endif.endif:
  %".264" = load i1, i1* %".231"
  %".265" = load i64, i64* %".234"
  br i1 %".264", label %"bb_body.endif.endif.endif.endif.endif.if", label %"bb_body.endif.endif.endif.endif.endif.endif", !prof !0
.239:
  br label %"bb_body.endif.endif.endif.endif.endif"
bb_body.endif.endif.endif.endif.if.if:
  call void @"PyErr_Clear"()
  %".242" = load {i8*, i32, i8*}, {i8*, i32, i8*}* %".229"
  %".243" = extractvalue {i8*, i32, i8*} %".242", 0
  %".244" = load {i8*, i32, i8*}, {i8*, i32, i8*}* %".229"
  %".245" = extractvalue {i8*, i32, i8*} %".244", 1
  %".246" = load {i8*, i32, i8*}, {i8*, i32, i8*}* %".229"
  %".247" = extractvalue {i8*, i32, i8*} %".246", 2
  %".248" = call i8* @"numba_unpickle"(i8* %".243", i32 %".245", i8* %".247")
  %".249" = icmp ne i8* null, %".248"
  br i1 %".249", label %"bb_body.endif.endif.endif.endif.if.if.if", label %"bb_body.endif.endif.endif.endif.if.if.endif", !prof !2
bb_body.endif.endif.endif.endif.if.endif:
  br i1 %".225", label %"bb_body.endif.endif.endif.endif.if.endif.if", label %"bb_body.endif.endif.endif.endif.if.endif.endif"
bb_body.endif.endif.endif.endif.if.if.if:
  call void @"numba_do_raise"(i8* %".248")
  br label %"bb_body.endif.endif.endif.endif.if.if.endif"
bb_body.endif.endif.endif.endif.if.if.endif:
  br label %".239"
bb_body.endif.endif.endif.endif.if.endif.if:
  call void @"PyErr_SetNone"(i8* @"PyExc_StopIteration")
  br label %".239"
bb_body.endif.endif.endif.endif.if.endif.endif:
  br i1 %".224", label %"bb_body.endif.endif.endif.endif.if.endif.endif.if", label %"bb_body.endif.endif.endif.endif.if.endif.endif.endif"
bb_body.endif.endif.endif.endif.if.endif.endif.if:
  br label %".239"
bb_body.endif.endif.endif.endif.if.endif.endif.endif:
  call void @"PyErr_SetString"(i8* @"PyExc_SystemError", i8* bitcast ([43 x i8]* @".const.unknown error when calling native function" to i8*))
  br label %".239"
bb_body.endif.endif.endif.endif.endif.if:
  ret i8* null
```
which looks a lot like the code that could be generated from here:
https://github.com/numba/numba/blob/18ab344052249b817919bd66cf0c24c066571709/numba/cpython/setobj.py#L497-L524

-----------
Use-case 4: returning python dict from objmode.
```
ret_type = types.DictType(types.unicode_type, types.int64)      # NOK (segfaults)
print(f"ret_type={ret_type}")
@njit
def test4():
    with objmode(res=ret_type):
        res = {'A': 1, 'B': 2}
    return res

result = test4()
print(result)
```

I don't think this is valid as Python dictionaries are not something Numba knows about and the return type is specified as a `typed.Dict` instance. This is equivalently invalid:
```python
from numba import njit, types, objmode
import numpy as np

ret_type = types.DictType(types.unicode_type, types.int64)      # NOK (segfaults)
print(f"ret_type={ret_type}")
@njit
def test4():
    with objmode(res=ret_type):
        res = "NOT A TYPED DICTIONARY"
    return res

result = test4()
print(result)
```
If however you returned a typed dictionary from the objmode block then I think that would be valid. e.g.:
```python
from numba import njit, types, objmode, typed
import numpy as np

ret_type = types.DictType(types.unicode_type, types.int64)
print(f"ret_type={ret_type}")
@njit
def test4():
    with objmode(res=ret_type):
        res = typed.Dict()
        res['A'] = 10
    return res

result = test4()
print(result)
```
I think some more type validation might be required somewhere, this sort of thing ideally shouldn't segfault!

**Comment by stuartarchibald** [comment]
The reported segfaults and incorrect behaviour are all fixed in Numba 0.53, closing.
